mathesar/utils/tables.py

```diff
@@ -90,13 +90,9 @@ def get_tables_meta_data(database_id):
     return TableMetaData.objects.filter(database__id=database_id)
 
 
-def patch_table_meta_data(table_oid, metadata_dict, database_id):
-    metadata_model = TableMetaData.objects.get(database__id=database_id, table_oid=table_oid)
-    alterable_fields = (
-        'import_verified', 'column_order', 'record_summary_customized', 'record_summary_template'
+def set_table_meta_data(table_oid, metadata, database_id):
+    TableMetaData.objects.update_or_create(
+        database=Database.objects.get(id=database_id),
+        table_oid=table_oid,
+        defaults={f: v for f, v in metadata.items() if f in table_metadata_fields}
```

Comment:
You can just pass the `metadata` dictionary directly. I can think of no cases where the behavior would be different. E.g., if you have extra keys in that dictionary, they'll be silently ignored.

Reply:
Apparently it's a bit more subtle than this. I'd still like to simplify this to match a more typical example. Given that the dictionary to be passed is documented and indeed typed, I want to avoid maintaining these fields in so many places.

Reply:
>  I can think of no cases where the behavior would be different. E.g., if you have extra keys in that dictionary, they'll be silently ignored.

Here are some cases:

- For an existing TableMetaData record, pass a `metadata` object like: 

    ```json
    {
      "table_oid": 99
    }
    ```

    The associated table is changed.

- For an existing TableMetaData record, pass a `metadata` object like: 

    ```json
    {
      "created_at": "1999-01-01 00:00:00"
    }
    ```

    The creation timestamp is changed.

- For a nonexistent TableMetaData record, pass a `metadata` object like:

    ```json
    {
      "id": 99
    }
    ```

    This creates a new record with the supplied id, messing up the PostgreSQL sequence, preventing any subsequent records from being created through the normal process

For stability's sake, I would prefer that the API not allow any of the above behavior.

I'd also like to point out that this field validation was already in place before my PR, right [here](https://github.com/mathesar-foundation/mathesar/blob/develop/mathesar/utils/tables.py#L95-L97). All I did was move the list of fields closer to the other code that it mirrors. The reason I moved it is precisely because I share your concern about that code getting out of sync. I actually hemmed and hawed about this for a while during this PR. I tried a played around with a few different approaches, basically kind of hating all of them. I'm finding that there is really an absurd amount of duplication and boilerplate with the patterns we're using. I don't blame us for this. I blame python. For example in `rpc/tables/metadata.py` there are _eight_ occurrences of `record_summary_customized`. To me that's crazy. It's crazy that we can't easily be more expressive about our intent with these fields.

I don't like having an additional list of metadata fields in `mathesar/models/base.py` either (as I originally had in this PR when I opened it). I tried two other approaches to maintaining the field validation while avoiding the need to maintain a list of fields.

- I tried finding a way to do some runtime introspection on the `TableMetaData` class, to get a list of field names set on that class but not its ancestors. And I wanted to then manually ignore fields like `database` and `table_oid`. That got kind of messy and started to bump up against the limits of my immediate knowledge of Python metaprogramming. I imagine this approach is possible. I just didn't want to continue spending time trying to figure out a clean way to do this.

- I also tried moving the `TableMetaDataBlob` TypedDict class lower down in the module graph. I tried putting it in `mathesar/models/base.py`. Then I imported it from there to `mathesar/utils/tables.py` so I could iterate over the fields for the sake of validation. But I couldn't figure out how to appropriate incorporate the _documentation_ of that class into the rendered output in MkDocs. I wrestled with this for quite a while actually. Probably at least 1 hour. I read over the [mkdocstrings docs](https://mkdocstrings.github.io/usage/) trying to figure out a way to do this. I imagine it's possible. But again, I didn't want to spend more time on it.

Ultimately I chose to prioritize application stability (validating the metadata fields) and development velocity (abandoning the above approaches) at the expense of code duplication. I still think that's the right balance to strike.

Given the counter examples I posed at the start of this comment, do you still maintain that we don't need any field validation here?

Do you have any better ideas for reducing code duplication?

Given that the list of fields was already in place before this PR, do you think that this code duplication is important enough to address in _this_ PR?

Reply:
We discussed this in our technical beta meeting today and decided to take Brent's approach for now but to hopefully come back to this topic later and find a better way to validate RPC request inputs in general.